File: tools/lib/bake.js
// Bakes a snippet body into the form the game receives.
//
// Three tools need this. `check-snippet-golden.js` locks the bytes a shipped build gets;
// `check-snippets.js` parses those same bytes; `check-state-eval.js` reads a baked body back to
// see that the press path measures each state the way the poll path does. They have to bake
// identically or one of them is guarding a body that never exists.
//
// Every stage comes out of `Snippets.lua` itself, run by `bake.lua`. Rewriting the rules in
// JavaScript would be a second copy that can drift, and a drifted copy silently guards something
// other than what gets baked.
const path = require("path");
const { runLuaDriver } = require("./lua51");

const driver = path.join(__dirname, "bake.lua");

/**
 * 주석 제거 → 프로브 치환. **`CONSTANTS.*` 치환은 안 한다** - 그 값은 `Constants`가 실제로
 * 실려 있어야 나오고, 부르는 쪽 둘 다 상수 값에는 관심이 없다. 검사는 그 자리를 문법상 올 수
 * 있는 것으로 메우고, 골든은 토큰 그대로 박제한다.
 *
 * 배포 갈래에서 값이 `false`인 프로브는 **호출이 통째로 사라진다.** 원문에서 멀쩡하던
 * `local x = PROBE.Winner(i)`가 `local x = `가 되는 자리라, 여기를 통과한 본문만 게임에서
 * 컴파일된다고 말할 수 있다.
 */
function bakeLive(body) {
    return runLuaDriver(driver, "live", body);
}

/** What the game actually receives: `BakeSnippet` itself, `CONSTANTS.*` turned into values. */
function bakeShipped(body) {
    return runLuaDriver(driver, "shipped", body);
}

/**
 * Reads one of `Constants`' string tables out as it stands.
 *
 * Folded onto tab and newline and fetched in one go. What is read here is single-line expressions,
 * so there is no fold to land inside a value.
 */
function constantStringTable(name) {
    const dump = runLuaDriver(driver, "constants", name);

    const out = {};
    if (dump.length === 0) return out;
    for (const row of dump.split("\n")) {
        const at = row.indexOf("\t");
        out[row.slice(0, at)] = row.slice(at + 1);
    }
    return out;
}

module.exports = { bakeLive, bakeShipped, constantStringTable };

File: tools/lib/bake.lua
-- Bakes a snippet body, or reads a constant table out, for the tools that check them.
--
-- `bake.js` runs this with the same lua5.1 the game and CI use, and hands the body across in a
-- file: bytes go out and come back untouched that way, where a pipe on Windows would translate
-- newlines and the golden is locked byte for byte.
--
--   lua5.1 tools/lib/bake.lua live|shipped <requestFile> <responseFile>
--   lua5.1 tools/lib/bake.lua constants <tableNameFile> <responseFile>

local mode, requestPath, responsePath = ...;
assert(mode and requestPath and responsePath, "usage: bake.lua <mode> <request> <response>");

local here = arg and arg[0] and arg[0]:match("^(.*)[/\\][^/\\]+$") or "tools/lib";
local srcDir = here .. "/../../Debind";

local function readFile(path)
    local file = assert(io.open(path, "rb"), path .. " could not be opened");
    local contents = file:read("*a");
    file:close();
    return contents;
end

local function writeFile(path, contents)
    local file = assert(io.open(path, "wb"));
    file:write(contents);
    file:close();
end

--- Runs one addon file the way the game does, handing it `local _, DebindPrivate = ...`.
local function runAddonFile(file, private)
    local chunk = assert(loadfile(srcDir .. "/" .. file));
    chunk("Debind", private);
end

--- What `Constants.lua` reads from the client while it loads.
---
--- **None of these is a rule of ours.** They are the client's own values, so a wrong one here is a
--- lie rather than a drift, and there is nothing in the repo to compare it against. Take them from
--- `reference/wow-ui-source/` when adding one.
local function installWowGlobals()
    _G.format = string.format;
    _G.tremove = table.remove;
    _G.MAX_PARTY_MEMBERS = 4;
    _G.MAX_RAID_MEMBERS = 40;
    _G.MAX_ARENA_ENEMIES = 5;
    _G.UnitClass = function() return "Warrior", "WARRIOR", 1; end
    -- The class table Constants.lua builds at load (CLASS_IDS). One real answer is enough: nothing
    -- baked reads that table, and what the loop needs is a call that does not raise and a range
    -- that ends.
    _G.C_CreatureInfo = {
        GetClassInfo = function(classID)
            if (classID == 1) then return { classFile = "WARRIOR" }; end
            return nil;
        end,
    };
end

--- `Snippets.lua` alone, on an empty `Constants`.
---
--- Nothing on this path looks at a constant's value, which is also why it stops short of
--- `BakeSnippet` -- that one does need the real ones.
local function loadSnippetsOnly()
    local private = { Constants = {} };
    runAddonFile("Snippets.lua", private);
    return private;
end

--- `Constants.lua` really loaded, so `BakeSnippet` itself can run, `CONSTANTS.*` included.
local function loadWithConstants()
    installWowGlobals();
    local private = {};
    -- The order is not free: `Snippets.lua` takes `DebindPrivate.Constants` as a local at load.
    runAddonFile("Constants.lua", private);
    runAddonFile("Snippets.lua", private);
    return private;
end

if (mode == "live") then
    local private = loadSnippetsOnly();
    local body = readFile(requestPath);
    writeFile(responsePath,
        private.applyProbesForTools(private.StripSnippetComments(body), private.SNIPPET_PROBES_LIVE));

elseif (mode == "shipped") then
    local private = loadWithConstants();
    writeFile(responsePath, private.BakeSnippet(readFile(requestPath)));

elseif (mode == "constants") then
    local private = loadWithConstants();
    local name = readFile(requestPath);
    local t = private.Constants[name];
    assert(t, "no such constant table: " .. name);
    local out = {};
    for k, v in pairs(t) do
        assert(type(v) == "string", name .. "." .. k .. " is not a string");
        out[#out + 1] = k .. string.char(9) .. v;
    end
    table.sort(out);
    writeFile(responsePath, table.concat(out, string.char(10)));

else
    error("unknown mode: " .. tostring(mode), 0);
end
